primitives/weights/src/weight_meter.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Contains the `WeightMeter` primitive to meter weight usage.
+
+use super::Weight;
+
+use sp_arithmetic::Perbill;
+
+/// Meters consumed weight and a hard limit for the maximal consumable weight.
+///
+/// Can be used to check if enough weight for an operation is available before committing to it.
+///
+/// # Example
+///
+/// ```rust
+/// use sp_weights::{Weight, WeightMeter};
+///
+/// // The weight is limited to (10, 0).
+/// let mut meter = WeightMeter::from_limit(Weight::from_parts(10, 0));
+/// // There is enough weight remaining for an operation with (5, 0) weight.
+/// assert!(meter.check_accrue(Weight::from_parts(5, 0)));
+/// // There is not enough weight remaining for an operation with (6, 0) weight.
+/// assert!(!meter.check_accrue(Weight::from_parts(6, 0)));
+/// ```
+#[derive(Debug, Clone)]
+pub struct WeightMeter {
+	/// The already consumed weight.
+	pub consumed: Weight,
+
+	/// The maximal consumable weight.
+	pub limit: Weight,
+}
+
+impl WeightMeter {
+	/// Creates [`Self`] from a limit for the maximal consumable weight.
+	pub fn from_limit(limit: Weight) -> Self {
+		Self { consumed: Weight::zero(), limit }
+	}
+
+	/// Creates [`Self`] with the maximal possible limit for the consumable weight.
+	pub fn max_limit() -> Self {
+		Self::from_limit(Weight::MAX)
+	}
+
+	/// The remaining weight that can be still consumed.
+	pub fn remaining(&self) -> Weight {
+		self.limit.saturating_sub(self.consumed)
+	}
+
+	/// The ratio of consumed weight to the limit.
+	///
+	/// Calculates one ratio per component and returns the largest.
+	pub fn remaining_ratio(&self) -> Perbill {
+		let time = Perbill::from_rational(self.consumed.ref_time(), self.limit.ref_time());
+		let pov = Perbill::from_rational(self.consumed.proof_size(), self.limit.proof_size());
+		time.max(pov)
+	}
+
+	/// Consume some weight and defensively fail if it is over the limit or would saturate the
+	/// arithmetic type.
+	pub fn defensive_accrue(&mut self, w: Weight) {
+		let res = self.check_accrue(w);
+		debug_assert!(res, "Weight meter exhausted; {} + {} > {}", &self.consumed, &w, &self.limit);
+	}
+
+	/// Check if the given weight can be consumed. Do nothing if not.
```

Comment:
```suggestion
	/// Consume the given weight after checking if it can be consumed. Do nothing if not.
```
…to emphasize the fact that this mutates the `consumed` weight.